src/Orleans/Providers/ILeaseProvider.cs

```diff
@@ -0,0 +1,38 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace Orleans.LeaseProviders
+{
+    /// <summary>
+    /// Aquired lease, which contains the resource key, duration of the lease and the token you can use to renew the lease with
+    /// </summary>
+    public class AcquiredLease
+    {
+        public string ResourceKey { get; }
+        public TimeSpan Duration { get; }
+        public string Token { get; }
+    }
+
+    /// <summary>
+    /// Lease request where you can specify ResourceKey, which you want to apply a lease on, and duration of your lease. 
+    /// </summary>
+    public class LeaseRequest
+    {
+        public string ResourceKey { get; set; }
+        public TimeSpan? Duration { get; set; }
+    }
+
+    /// <summary>
+    /// Lease provider interface 
+    /// </summary>
+    public interface ILeaseProvider
```

Comment:
It's a worthwhile question to raise. This lease manager does not currently need to know the total set resources it manages. If we had an API which queried which leases are free, it would need some way of enumerating them. Would that API return a consistent view of the leases or would it query one-by-one and hence return a torn view? Maybe the answer is the latter (which is fine), but questions like that would need to be answered.

I'm not strongly against it, but just wanted to make sure that we determined it was worth it and that we've considered the nuances.